### PR TITLE
RGW - Zipper 9: Write path

### PR DESCRIPTION
**Pull request by dang**
This series pulls the write path into Zipper.  In addition, it converts a few misc. things like lifecycle, swift versioning, and serializer.

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by dang** [comment]
https://pulpito.ceph.com/dang-2020-11-19_15:20:29-rgw-wip-dang-zipper-9-distro-basic-smithi/

File: src/rgw/rgw_cr_rados.cc [hunk 1]
```diff
@@ -635,7 +635,7 @@ int RGWAsyncFetchRemoteObj::_send_request()
 
   char buf[16];
   snprintf(buf, sizeof(buf), ".%lld", (long long)store->getRados()->instance_id());
```
Comment (mattbenjamin) [inline src/rgw/rgw_cr_rados.cc:637]: this isn't new, but should we use the explicit integral notations in places like this, and also explicit integral types?

File: src/rgw/rgw_sal.h [hunk 2]
```diff
@@ -415,6 +448,7 @@ class RGWObject {
 
     /* OPs */
     virtual std::unique_ptr<ReadOp> get_read_op(RGWObjectCtx *) = 0;
+    virtual std::unique_ptr<WriteOp> get_write_op(RGWObjectCtx *) = 0;
```
Comment (mattbenjamin) [inline src/rgw/rgw_sal.h:512]: RGWObjectCtx*

File: src/rgw/rgw_sal_rados.cc [hunk 3]
```diff
@@ -694,6 +702,56 @@ int RGWRadosObject::RadosReadOp::iterate(int64_t ofs, int64_t end, RGWGetDataCB
   return parent_op.iterate(ofs, end, cb, y);
 }
 
+std::unique_ptr<RGWObject::WriteOp> RGWRadosObject::get_write_op(RGWObjectCtx *ctx)
```
Comment (mattbenjamin) [inline src/rgw/rgw_sal_rados.cc:771]: RGWObjectCtx*

File: src/rgw/rgw_sal_rados.cc [hunk 4]
```diff
@@ -694,6 +702,56 @@ int RGWRadosObject::RadosReadOp::iterate(int64_t ofs, int64_t end, RGWGetDataCB
...
+{
+  return std::unique_ptr<RGWObject::WriteOp>(new RGWRadosObject::RadosWriteOp(this, ctx));
+}
+
+RGWRadosObject::RadosWriteOp::RadosWriteOp(RGWRadosObject *_source, RGWObjectCtx *_rctx) :
```
Comment (mattbenjamin) [inline src/rgw/rgw_sal_rados.cc:776]: RGWRadosObject*

File: src/rgw/rgw_sal_rados.h [hunk 5]
```diff
@@ -67,6 +67,21 @@ class RGWRadosObject : public RGWObject {
...
 
+    struct RadosWriteOp : public WriteOp {
+    private:
+      RGWRadosObject* source;
+      RGWObjectCtx* rctx;
+      RGWRados::Object op_target;
+      RGWRados::Object::Write parent_op;
+
+    public:
+      RadosWriteOp(RGWRadosObject *_source, RGWObjectCtx *_rctx);
```
Comment (mattbenjamin) [inline src/rgw/rgw_sal_rados.h:79]: RGWObjectCtx*

File: src/rgw/rgw_lc.cc [hunk 6]
```diff
@@ -556,9 +545,11 @@ struct lc_op_ctx {
 	    const DoutPrefixProvider *dpp, WorkQ* wq)
     : cct(env.store->ctx()), env(env), o(o), next_key_name(next_key_name),
       effective_mtime(effective_mtime),
-      store(env.store), bucket_info(env.bucket_info), op(env.op), ol(env.ol),
-      obj(env.bucket_info.bucket, o.key), rctx(env.store), dpp(dpp), wq(wq)
-    {}
+      store(env.store), bucket(env.bucket), op(env.op), ol(env.ol),
+      rctx(env.store), dpp(dpp), wq(wq)
+    {
+    obj = bucket->get_object(o.key);
```
Comment (mattbenjamin) [inline src/rgw/rgw_lc.cc:551]: indent?

File: src/rgw/rgw_lc.cc [hunk 7]
```diff
@@ -1610,30 +1598,28 @@ int RGWLC::bucket_lc_post(int index, int max_lock_sec,
...
-    ret = cls_rgw_lc_set_entry(store->getRados()->lc_pool_ctx,
-			       obj_names[index],  entry);
+    ret = sal_lc->set_entry(obj_names[index],  entry);
     if (ret < 0) {
       ldpp_dout(this, 0) << "RGWLC::process() failed to set entry on "
           << obj_names[index] << dendl;
     }
 clean:
-    l.unlock(&store->getRados()->lc_pool_ctx, obj_names[index]);
+    lock->unlock();
```
Comment (mattbenjamin) [inline src/rgw/rgw_lc.cc:1607]: this also isn't new, but could this have used a unique_lock?

File: src/rgw/rgw_lc.cc [hunk 8]
```diff
@@ -1718,19 +1704,19 @@ int RGWLC::process(int index, int max_lock_secs, LCWorker* worker,
 	  << "index: " << index << " worker ix: " << worker->ix
 	  << dendl;
 
-  rados::cls::lock::Lock l(lc_index_lock_name);
+  rgw::sal::LCSerializer* lock = sal_lc->get_serializer(lc_index_lock_name,
```
Comment (mattbenjamin) [inline src/rgw/rgw_lc.cc:1707]: oh, we're getting a pointer;  still...

File: src/rgw/rgw_lc.cc [hunk 9]
```diff
@@ -502,8 +491,8 @@ class LCObjsLister {
   }
 
   boost::optional<std::string> next_key_name() {
-    if (obj_iter == objs.end() ||
-	(obj_iter + 1) == objs.end()) {
+    if (obj_iter == list_results.objs.end() ||
+	(obj_iter + 1) == list_results.objs.end()) {
```
Comment (mattbenjamin) [inline src/rgw/rgw_lc.cc:495]: the next_key_name() mechanism needs to still work--I think I wrote improved s3-tests cases for this, but we should be sure

File: src/rgw/rgw_obj_manifest.h [hunk 10]
```diff
@@ -44,7 +45,7 @@ class rgw_obj_select {
   }
 
   rgw_raw_obj get_raw_obj(const RGWZoneGroup& zonegroup, const RGWZoneParams& zone_params) const;
-  rgw_raw_obj get_raw_obj(RGWRados *store) const;
+  rgw_raw_obj get_raw_obj(rgw::sal::RGWStore *store) const;
```
Comment (mattbenjamin) [inline src/rgw/rgw_obj_manifest.h:48]: rgw::sal::RGWStore*

File: src/rgw/rgw_rados.cc [hunk 11]
```diff
@@ -149,11 +149,11 @@ rgw_raw_obj rgw_obj_select::get_raw_obj(const RGWZoneGroup& zonegroup, const RGW
   return raw_obj;
 }
 
-rgw_raw_obj rgw_obj_select::get_raw_obj(RGWRados *store) const
+rgw_raw_obj rgw_obj_select::get_raw_obj(rgw::sal::RGWStore *store) const
```
Comment (mattbenjamin) [inline src/rgw/rgw_rados.cc:152]: rgw::sal::RGWStore*

File: src/rgw/rgw_sal.h [hunk 12]
```diff
@@ -455,7 +460,22 @@ class RGWObject {
...
     virtual MPSerializer* get_serializer(const std::string& lock_name) = 0;
+    virtual int transition(RGWObjectCtx& rctx,
+			   RGWBucket* bucket,
+			   const rgw_placement_rule& placement_rule,
+			   const real_time& mtime,
+			   uint64_t olh_epoch,
+			   const DoutPrefixProvider *dpp,
+			   optional_yield y) = 0;
+    virtual int get_max_chunk_size(rgw_placement_rule placement_rule,
+				   uint64_t *max_chunk_size,
```
Comment (mattbenjamin) [inline src/rgw/rgw_sal.h:475]: uint64_t*

**Review by mattbenjamin — APPROVED** [review]
this looks very good, and teuthology run seemed to ass
running s3-tests for lifecycle by hand would be appropriate

File: src/rgw/rgw_lc.cc (hunk 8 above)
Comment (dang) [reply src/rgw/rgw_lc.cc:1707]: Yeah, it's because of inheritance.

File: src/rgw/rgw_lc.cc (hunk 9 above)
Comment (dang) [reply src/rgw/rgw_lc.cc:495]: It should be good.

**Review by cbodley — APPROVED** [review]

**Comment by cbodley** [comment]
has @adamemerson looked over the static_ptr changes?

File: src/common/static_ptr.h [hunk 13]
```diff
@@ -33,23 +33,14 @@ namespace _mem {
...
   auto me = static_cast<T*>(p1);
 
   switch (oper) {
-  case op::copy:
-    // One conspicuous downside is that immovable/uncopyable functions
-    // kill compilation right here, even if nobody ever calls the move
-    // or copy methods. Working around this is a pain, since we'd need
-    // four operator functions and a top-level class to
-    // provide/withhold copy/move operations as appropriate.
-    new (p2) T(*me);
```
Comment (adamemerson) [inline src/common/static_ptr.h:50]: I approve.

**Review by adamemerson — APPROVED** [review]
Looks good tome.

**Pull request merged**